doc/compiler/optimizer/inlineFastPathLocations.md

```diff
@@ -0,0 +1,119 @@
+recognized methods into `RecognizedCallTransformer`. The majority of them have not been done. Only a few methods are moved
+to `RecognizedCallTransformer` to avoid the complexities of doing unsafe transformations in the middle of an already
+complex transformation to inline calls. It also has the added benefit of avoiding falling victim to the inliner heuristics
+and somehow end up not inlining the call.
+
+## Inlining Fast Path Optimization In ILGen
+
+ILGen does not operate on trees, but it operates on bytecodes. It produces trees. It operates on the same operand stack
+in Java bytecodes. It must track how many things are on the stack. It is not a very intuitive format to work with for
+transformations. It is easier to do the transformation on well-formed trees rather than during the trees are being generated.
+
+ILGen should be kept as simple as possible. The task of optimizations should not be mixed with the task of generating IL.
+After IL is generated, `ilgenStrategyOpts` will run before other optimizations see the IL. To avoid the complexity in the
+ILGen and to keep things simpler, the recognized methods in IlGen could be moved to `RecognizedCallTransformer` whose
+purpose is to transform recognized calls.
+
+## UnsafeFastPath
+
+`UnsafeFastPath` focuses on `sun.misc.Unsafe` methods. `sun.misc.Unsafe` methods allow low level programming. For example,
+`sun.misc.Unsafe.getByte(Object o, long offset)` takes the address of an object and adds it with the offset and reads a
+byte out of it. Even if the object `o` might not be a byte array or might not even be an array, you could still read a
+byte from the address. Unsafe calls allow to read any memory.
+
+Another example is that we can dereference an object to get a `J9Class` pointer. The `J9Class` pointer can be dereferenced
+into a class property pointer which can be dereferenced again. All these kinds of dereferences cannot be written in normal
+Java code because they go through the JVM native data structures. They can be written in unsafe code. Unsafe path allows
+us to access memory without any safe type checking or bound checking, etc. It also allows us to look at native memory
+(`sun.misc.Unsafe.allocateMemory()`).
+
+Permission change is required to use unsafe APIs. It is not meant to be used in general programing. It is meant to be a
+tool to be used internally within JCL. Unsafe inlining is important because unsafe methods are used often by JCL. Calling
+into natives is slow. It is a lot faster to use JIT inlined expressions.
+
+`UnsafeFastPath` reduces unsafe calls to a "no control flow" sequence of IL trees to do the required operation. For example,
+`getInt(Object o, long offset)` adds the address of an object `o` and adds it with the offset and loads an integer. They
+can be expressed to a "no control flow" sequence of IL trees to do the required operation directly (e.g. `load` or `store`).
+
```

Comment:
The last sentence may be redundant given what was said earlier

Reply:
I simplified the last sentence as
> They can be expressed as the required operations directly (e.g. `load` or `store`)